src/win32/path_w32.c

```diff
@@ -303,3 +308,230 @@ char *git_win32_path_8dot3_name(const char *path)
+						if (git_win32_path_readlink_w(target, dir->f.cFileName) >= 0) {
+							ps->st.st_mode = (ps->st.st_mode & ~S_IFMT) | S_IFLNK;
+
+							/* st_size gets the UTF-8 length of the target name, in bytes,
+							*  not counting the NULL terminator */
+							if ((ps->st.st_size = git__utf16_to_8(NULL, 0, target)) < 0) {
+								error = -4;
+								giterr_set(GITERR_OS, "Could not manage reparse link '%s'", dir->f.cFileName);
+								goto clean_up_and_exit;
+							}
+						}
+					}
+
+					if (S_ISDIR(ps->st.st_mode)) {
+						ps->path[ps->path_len++] = '/';
+						ps->path[ps->path_len] = '\0';
+					} else if (!S_ISREG(ps->st.st_mode) && !S_ISLNK(ps->st.st_mode)) {
+						git__free(ps);
+						ps = NULL;
+					}
+
+					if (ps)
+					{
+						git_vector_insert(contents, ps);
+					}
+				}
+			}
+		}
+
+		memset(&dir->f, 0, sizeof(git_path_with_stat));
+		dir->first = 0;
+
+		if (!FindNextFileW(dir->h, &dir->f))
+		{
+			if (GetLastError() == ERROR_NO_MORE_FILES)
+				break;
+			else {
+				error = -5;
+				giterr_set(GITERR_OS, "Could not get attributes for file in '%s'", path);
+				goto clean_up_and_exit;
+			}
+		}
+	}
+
+	/* sort now that directory suffix is added */
+	git_vector_sort(contents);
+
+clean_up_and_exit:
+
+	if (dir) {
+		FindClose(dir->h);
+		free(dir);
+	}
+
+	return error;
+}
+#endif
+
+static bool path_is_volume(wchar_t *target, size_t target_len)
+{
+	return (target_len && wcsncmp(target, L"\\??\\Volume{", 11) == 0);
+}
+
+/* On success, returns the length, in characters, of the path stored in dest.
+* On failure, returns a negative value. */
+int git_win32_path_readlink_w(git_win32_path dest, const git_win32_path path)
+{
+	BYTE buf[MAXIMUM_REPARSE_DATA_BUFFER_SIZE];
+	GIT_REPARSE_DATA_BUFFER *reparse_buf = (GIT_REPARSE_DATA_BUFFER *)buf;
+	HANDLE handle = NULL;
+	DWORD ioctl_ret;
+	wchar_t *target;
+	size_t target_len;
+
+	int error = -1;
+
+	handle = CreateFileW(path, GENERIC_READ,
+		FILE_SHARE_READ | FILE_SHARE_DELETE, NULL, OPEN_EXISTING,
+		FILE_FLAG_OPEN_REPARSE_POINT | FILE_FLAG_BACKUP_SEMANTICS, NULL);
+
+	if (handle == INVALID_HANDLE_VALUE) {
+		errno = ENOENT;
+		return -1;
+	}
+
+	if (!DeviceIoControl(handle, FSCTL_GET_REPARSE_POINT, NULL, 0,
+		reparse_buf, sizeof(buf), &ioctl_ret, NULL)) {
+		errno = EINVAL;
+		goto on_error;
+	}
+
+	switch (reparse_buf->ReparseTag) {
+	case IO_REPARSE_TAG_SYMLINK:
+	target = reparse_buf->SymbolicLinkReparseBuffer.PathBuffer +
+		(reparse_buf->SymbolicLinkReparseBuffer.SubstituteNameOffset / sizeof(WCHAR));
+	target_len = reparse_buf->SymbolicLinkReparseBuffer.SubstituteNameLength / sizeof(WCHAR);
+	break;
+	case IO_REPARSE_TAG_MOUNT_POINT:
+	target = reparse_buf->MountPointReparseBuffer.PathBuffer +
+		(reparse_buf->MountPointReparseBuffer.SubstituteNameOffset / sizeof(WCHAR));
+	target_len = reparse_buf->MountPointReparseBuffer.SubstituteNameLength / sizeof(WCHAR);
+	break;
+	default:
+	errno = EINVAL;
```

Comment:
Thanks for these. For whatever reason I was having an issue seeing the style problems. Looks like I somehow mucked up the function's formatting when I moved it from posix_w32.c to path_w32.c.